components/net/http_cache.rs

```diff
@@ -657,10 +676,21 @@ impl HttpCache {
         if let ResponseBody::Done(ref completed_body) = *response.body.lock().unwrap() {
             let entry_key = CacheKey::new(request.clone());
             if let Some(cached_resources) = self.entries.get(&entry_key) {
-                for cached_resource in cached_resources.iter() {
+                // Ensure we only wake-up consumers of relevant resources,
+                // ie we don't want to wake-up 200 awaiting consumers with a 206.
+                let relevant_cached_resources = cached_resources
+                    .iter()
+                    .filter(|resource| resource.data.raw_status == response.raw_status);
+                for cached_resource in relevant_cached_resources {
                     let mut awaiting_consumers = cached_resource.awaiting_body.lock().unwrap();
                     for done_sender in awaiting_consumers.drain(..) {
-                        if cached_resource.aborted.load(Ordering::Relaxed) {
+                        if cached_resource.aborted.load(Ordering::Relaxed) ||
+                            response.is_network_error()
```

Comment:
Will we ever reach this check if it's a network error? Won't the [raw_status value](https://github.com/servo/servo/blob/master/components/net_traits/response.rs#L162) not match the ones in cached_resources, or do we actually store network error responses in the cache?

Reply:
Since we immediately store responses when the networking starts, we do store responses that later error, which in this case is similar to the corresponding fetch having been aborted. I think the fact that we didn't wake-up readers awaiting a response that by then had error-ed, was a bug. 

And we also store network errors that are defined as "cacheable by default", see https://tools.ietf.org/html/rfc7231#section-6.1